forum/search/typesense.py

```diff
@@ -0,0 +1,423 @@
+"""
+Typesense backend for searching comments and threads.
+"""
+
+from typing import Any, Optional
+
+from bs4 import BeautifulSoup
+from django.conf import settings
+from django.core.paginator import Paginator
+
+# mypy complains: 'Module "typesense" does not explicitly export attribute "Client"'
+from typesense import Client  # type: ignore
+from typesense.types.collection import CollectionCreateSchema
+from typesense.types.document import DocumentSchema, SearchParameters
+from typesense.exceptions import ObjectNotFound
+
+from forum.backends.mysql.models import Comment, CommentThread
+from forum.constants import FORUM_MAX_DEEP_SEARCH_COMMENT_COUNT
+from forum.search.base import (
+    BaseDocumentSearchBackend,
+    BaseIndexSearchBackend,
+    BaseSearchBackend,
+    BaseThreadSearchBackend,
+)
+
+_TYPESENSE_CLIENT: Client | None = None
+
+
+def get_typesense_client() -> Client:
+    """
+    Return a singleton Typesense client instance.
+    """
+    global _TYPESENSE_CLIENT
+    if _TYPESENSE_CLIENT is None:
+        _TYPESENSE_CLIENT = Client(
+            {
+                "api_key": settings.TYPESENSE_API_KEY,
+                "nodes": settings.TYPESENSE_URLS,
+            }
+        )
+    return _TYPESENSE_CLIENT
+
+
+def quote_filter_value(value: str) -> str:
+    """
+    Sanitize and safely quote a value for use in a Typesense filter.
+
+    https://typesense.org/docs/guide/tips-for-filtering.html#escaping-special-characters
+    """
+    return "`" + value.replace("`", "") + "`"
+
+
+def collection_name() -> str:
+    """
+    Generate the collection name to use in Typesense.
+    """
+    return settings.TYPESENSE_COLLECTION_PREFIX + "forum"
+
+
+def collection_schema() -> CollectionCreateSchema:
+    """
+    The schema to use for creating the collection.
+    """
+    return {
+        "name": collection_name(),
+        # NOTE: there's always an implicit "id" field
+        "fields": [
+            {"name": "thread_id", "type": "string"},
+            {"name": "course_id", "type": "string"},
+            {"name": "commentable_id", "type": "string"},
+            {"name": "context", "type": "string"},
+            {"name": "text", "type": "string"},
+        ],
+    }
+
+
+def expected_full_collection_schema() -> dict[str, Any]:
+    """
+    What is expected to be the full collection schema.
+
+    Use this to validate the actual schema from the server.
+    """
+    return {
+        "default_sorting_field": "",
+        "enable_nested_fields": False,
+        "fields": [
+            {
+                "facet": False,
+                "index": True,
+                "infix": False,
+                "locale": "",
+                "name": "thread_id",
+                "optional": False,
+                "sort": False,
+                "stem": False,
+                "stem_dictionary": "",
+                "store": True,
+                "type": "string",
+            },
```

Comment:
Maybe factor out all the common default values here, so each of these becomes something much more readable like `{ **defaults, name: "thread_id" }` ?